lib/src/fixed_datetime_formatter.dart

```diff
@@ -0,0 +1,306 @@
+      if (previousEnd < start) {
+        buffer.write(pattern.substring(previousEnd, start));
+      }
+      var formatCharacter = _blocks.formatCharacters[i];
+      var number = _extractNumFromDateTime(
+        formatCharacter,
+        datetime,
+      ).toString();
+      if (number.length > length) {
+        if (formatCharacter == _fractionSecondCode) {
+          //Special case, as we want fractional seconds to be the leading digits
+          number = number.substring(length);
+        } else {
+          number = number.substring(number.length - length);
+        }
+      } else if (number.length < length) {
+        number = number.padLeft(length, '0');
+      }
+      buffer.write(number);
+    }
+    if (_blocks.length > 0) {
+      var lastEnd = _blocks.ends.last;
+      if (lastEnd < pattern.length) {
+        buffer.write(pattern.substring(lastEnd, pattern.length));
+      }
+    }
+    return buffer.toString();
+  }
+
+  int _extractNumFromDateTime(int? formatChar, DateTime dateTime) {
+    switch (formatChar) {
+      case _yearCode:
+        return dateTime.year;
+      case _centuryCode:
+        return (dateTime.year / 100).floor();
+      case _decadeCode:
+        return (dateTime.year / 10).floor();
+      case _monthCode:
+        return dateTime.month;
+      case _dayCode:
+        return dateTime.day;
+      case _hourCode:
+        return dateTime.hour;
+      case _minuteCode:
+        return dateTime.minute;
+      case _secondCode:
+        return dateTime.second;
+      case _fractionSecondCode:
+        return dateTime.microsecond;
+    }
+    throw AssertionError("Unreachable, the key is checked in the constructor");
+  }
+
+  /// Parse a string [formattedDateTime] to a local [DateTime] as specified in the
+  /// [pattern], substituting missing values with a default. Throws an exception
+  /// on failure to parse.
+  DateTime decode(String formattedDateTime) {
+    return _decode(formattedDateTime, isUtc, true);
+  }
+
+  /// Same as [decode], but will not throw on parsing erros, instead using
+  /// the default value as if the format char was not present in the [pattern].
+  DateTime tryDecode(String formattedDateTime) {
+    return _decode(formattedDateTime, isUtc, false);
+  }
+
+  DateTime _decode(
+    String formattedDateTime,
+    bool isUtc,
+    bool throwOnError,
+  ) {
+    var characters = formattedDateTime.codeUnits;
+    var year = 0;
+    var century = 0;
+    var decade = 0;
+    var month = 1;
+    var day = 1;
+    var hour = 0;
+    var minute = 0;
+    var second = 0;
+    var microsecond = 0;
+    for (int i = 0; i < _blocks.length; i++) {
+      var char = _blocks.formatCharacters[i];
+      var number = _extractNumFromString(characters, i, throwOnError);
+      if (number != null) {
+        if (char == _fractionSecondCode) {
+          //Special case, as we want fractional seconds to be the leading digits
+          var numberLength = _blocks.ends[i] - _blocks.starts[i];
```

Comment:
Needs to account for `numberLength` > 6 too. (Cut off the low digits).

Maybe:
```dart
if (numberLength > 6) {
 while (numberLength >= 12) { // Probably rare.
   number ~/ 1000000;
   numberLength -= 6;
 }
 number ~/ _powersOfTen[numberLength - 6];
} else if (numberLength < 6) {
  number *= _powersOfTen[6 - numberLength];
} 
```

(This also raises the problem of someone using `"YYYYYYYYYYYYYYYYYYYY"` as format, and the number parser overflowing. Maybe just put a limit on the length of fields to, say, 10 format characters. Anything more, and we throw when parsing the pattern. That should ensure that you can parse any *reasonable* value, and not have to care about overflows.)

Reply:
As I wrote above, I would rather throw an exception here. Cutting off digits is not expected behavior, so it might lead to situations where a user wants to parse nanoseconds, gets no error but wrong `DateTime`s. 

Concerning the overflow: As per its documentation, 
```
DateTimes can represent time values that are at a distance of at most 100,000,000 days from epoch (1970-01-01 UTC): -271821-04-20 to 275760-09-13.
```
As [DateTime] does not throw an error when constructing dates beyond that, I would not do it here either to make the behavior consistent.

Reply:
DateTime does throw when you exceed the +/-100000000 days range: 
```dart
void main() {
  print(DateTime.utc(1970, 1, 100000001, 0, 0, 0)); // works
  print(DateTime.utc(1970, 1, 100000001, 0, 0, 1)); // throws
}
```
If we could rely on that catching the errors, that would be fine.

I guess six digits is enough for fractional seconds. If you get inputs with more digits,
you can't use the extra digits anyway, and you can make the format be `"ss.SSSSSS000000"` to ignore the later digits.


We still need to handle larger field sizes on the other, "normal", cases.

Consider a format of `"YYYYYYYYYYYYYYYYYYYY"` (20 digits).

You can *encode* to that without issue, and you can decode if the value is small enough,
but you can also run into inputs which overflow the parsing logic, like
`"36893488147419103233"` (20 digits, value is 2^65+1).
The parsing logic relies on `value = value * 10 + digit;` in a loop, and on the VM that integer will overflow and give you a result of `1`.
There are other inputs which can result in a *negative* value (which is probably not any more problematic.)

That's why I want to restrict the length of format fields, so that the number _parsing_ cannot overflow. A limit of 19 is enough for the VM, but a limit of 9 should be enough for any reasonable value, and will keep the computed number in the Smi range on 32-bit platforms too.

Reply:
I think it's kind of weird how the boundaries are enforced in `DateTime`...
```
void main() {
  print(DateTime.utc(1970, 1, 100000001, 0, 0, 0)); // works
  print(DateTime.utc(1970, 1, 100000001, 0, 0, 1)); // throws
  // but
  print(DateTime.utc(100000001, 1, 100000000, 0, 0, 1)); // works
}
```
As it makes sense to prevent overflows, and as there is no sane reason I can think of to have *that* many characters in a format block, I implemented the limit of 9 which should be more than sufficient.